### DISCUSSION TITLE
How to cyclize a molecule, e.g. a linear peptide, into a cyclic one?

### DISCUSSION THREAD

**Discussion by charlesxu90** [rdkit/rdkit#6255] [Q&A] [answered]
Dear rdkit experts @greglandrum @ptosco,

I'm Charles XU, and I'm trying to design cyclic peptides using RDKit. I met a problem during cyclizing of a peptide. 

My initial notation is based on HELM, and I was able to create a linear peptide following the orders, as shown below. However, I didn't find any useful way to cyclize the peptide by connecting the '_R3' and '_R2' groups together using RDKit.

- Original monomers:
```['CC([*:_R2])=O', '[*:_R1]N1CCC[C@H]1C([*:_R2])=O', 'CC(C)C[C@@H](N(C)[*:_R1])C([*:_R2])=O', 'C[C@@H](O[*:_R3])[C@H](N[*:_R1])C([*:_R2])=O', 'CC(C)C[C@H](N(C)[*:_R1])C([*:_R2])=O', 'CC(C)C[C@H](N[*:_R1])C([*:_R2])=O', 'CC(C)C[C@H](N[*:_R1])C([*:_R2])=O', 'C[C@H](N(C)[*:_R1])C([*:_R2])=O', 'CC(C)C[C@@H](N[*:_R1])C([*:_R2])=O', '[*:_R1]N1CCC[C@@H]1C([*:_R2])=O']```

- Linear peptide:
```'[*:_R3]O[C@H](C)[C@H](NC(=O)[C@@H](CC(C)C)N(C)C(=O)[C@@H]1CCCN1C(C)=O)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H],(CC(C)C)C(=O)N(C)[C@@H](C)C(=O)N[C@H](CC(C)C)C(=O)N1CCC[C@@H]1C([*:_R2])=O'```

- Cyclic link:
```'4:R3-10:R2'```

One intuitive way is to use `Chem.molzip` in RDKit, but it didn't work for single molecules. Alternatively, I thought of creating a reaction and using it for the conversion. However, I found the reaction would be different for different peptides, so the conversion process cannot be standardized well.

Any suggestions and answers would be highly appreciated!

Thanks and best regards,

Charles XU

**Comment by charlesxu90** [comment]
I tried to replace the `_R2` and `_R3` into atom maps `1` the linear peptide and run `Chem.molzip`, but it didn't work.

`'[*:1]O[C@H](C)[C@H](NC(=O)[C@@H](CC(C)C)N(C)C(=O)[C@@H]1CCCN1C(C)=O)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H],(CC(C)C)C(=O)N(C)[C@@H](C)C(=O)N[C@H](CC(C)C)C(=O)N1CCC[C@@H]1C([*:1])=O'`

**Comment by charlesxu90** [comment]
Using an reaction to conduct the connectiong relies on the atoms nearby. I don't think there would be an intuitive way to connect these two atoms in different peptides using a generation reaction.

**Reply by greglandrum** [reply]
@charlesxu90 there may not be a straightforward way to do this, but it’s worth trying.
Can you please provide one specific input example and exactly what you would like to get as output?

**Comment by charlesxu90** [comment]
@greglandrum Below is the original input HELM sequence:
`'{[ac-].P.[Me_dL].T.[meL].L.L.[meA].[dL].[dP]}', '4:R3-10:R2'`

The monomers are in CXSMILES as shown below:

```text
ID	Symbol	Compound_Name	Natural_Analog	replaced_SMILES	CXSMILES	Monomer_Type	R1	R2	R3

3	meA	N-methyl-L-alanine	A	CN[C@@H](C)C(=O)O	C[C@H](N(C)[*])C([*])=O |$;;;;_R1;;_R2;$|	Backbone	H	OH	-
100	L	Leucine	L	CC(C)C[C@H](N)C(=O)O	CC(C)C[C@H](N[*])C([*])=O |$;;;;;;_R1;;_R2;$|	Backbone	H	OH	-
101	dL	D-leucine	L	CC(C)C[C@@H](N)C(=O)O	CC(C)C[C@@H](N[*])C([*])=O |$;;;;;;_R1;;_R2;$|	Backbone	H	OH	-
102	meL	N-methyl-L-leucine	L	CN[C@@H](CC(C)C)C(=O)O	CC(C)C[C@H](N(C)[*])C([*])=O |$;;;;;;;_R1;;_R2;$|	Backbone	H	OH	-
103	Me_dL	N-methyl-D-leucine	L	CN[C@H](CC(C)C)C(=O)O	CC(C)C[C@@H](N(C)[*])C([*])=O |$;;;;;;;_R1;;_R...	Backbone	H	OH	-
129	P	Proline	P	O=C(O)[C@@H]1CCCN1	[*]N1CCC[C@H]1C([*])=O |$_R1;;;;;;;_R2;$|	Backbone	H	OH	-
130	dP	D-proline	P	O=C(O)[C@H]1CCCN1	[*]N1CCC[C@@H]1C([*])=O |$_R1;;;;;;;_R2;$|	Backbone	H	OH	-
166	T	Threonine	T	C[C@@H](O)[C@H](N)C(=O)O	C[C@@H](O[*])[C@H](N[*])C([*])=O |$;;;_R3;;;_R...	Backbone	H	OH	H
306	ac-	N-Terminal acetic acid	X	CC(=O)O	CC([*])=O |$;;_R2;$|	Terminal	-	OH	-
```

I first linked the monomers into a linear peptide following [this gist](https://gist.github.com/iwatobipen/1ffa9b2a13b70cf3f424e7dd9af68a90). The results is:
 `'[*:_R3]O[C@H](C)[C@H](NC(=O)[C@@H](CC(C)C)N(C)C(=O)[C@@H]1CCCN1C(C)=O)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H],(CC(C)C)C(=O)N(C)[C@@H](C)C(=O)N[C@H](CC(C)C)C(=O)N1CCC[C@@H]1C([*:_R2])=O'
`
Where `'[*:_R3]` and `[*:_R2]` has to be connected. But I have no clear clue to do that yet.

**Comment by charlesxu90** [comment]
Below is a function that I wrote for this, but it doesn't work as expected as `Chem.molzip` cannot do this connection.
```python
def cyclize_peptide(smi, link):
    """
    Cyclize a peptide through a link, through definiting a user-defined reaction
    smi: SMILES of a linear peptide, e.g. '[*:_R3]O[C@H](C)[C@H](NC(=O)[C@@H](CC(C)C)N(C)C(=O)[C@@H]1CCCN1C(C)=O)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N(C)[C@@H](C)C(=O)N[C@H](CC(C)C)C(=O)N1CCC[C@@H]1C([*:_R2])=O'
    link: the link between the two ends, e.g. '4:R3-10:R2'
    """
    mol = Chem.MolFromSmiles(get_cxsmiles_from_smi(smi))
    display(mol)
    link = link.split('-')
    start, end = link[0].split(':')[1], link[1].split(':')[1]
    
    for atm in mol.GetAtoms():
        if atm.HasProp("atomLabel") and atm.GetProp("atomLabel") in [start, end]:
            atm.SetAtomMapNum(10)  # Use 10 as the atom map number
    mol = Chem.molzip(mol)
    display(mol)
    cxsmi = Chem.MolToCXSmiles(mol, isomericSmiles=True)
    smi = get_smi_from_cxsmiles(clean_dummy_labels_in_cxsmiles(cxsmi))
    return smi
```

The three util functions are as below:
```python
# Get CXSMILES from SMILES
def get_cxsmiles_from_smi(smi):
    """
    Get CXSMILES from SMILES
    Input: 'CCCC[C@H](N(C)[*:_R1])C([*:_R2])=O'
    Output: 'CCCC[C@H](N(C)[*])C([*])=O|$;;;;;;;_R1;;_R2;$|'
    """
    cxsmiles = smi
    labels = re.findall(r'\[\*\:(.*?)\]', smi)  # Get all labels with pattern [*:label]
    r_groups = []
    for label in labels:
        cxsmiles = cxsmiles.replace(f'[*:{label}]', '[*]') # Replace [*:label] with [*]
        r_groups.append(f'{label}')     # Record label to r_groups list
    
    pos = list()
    r_group_idx = 0
    # Iterate through the SMILES and add ';'s in pos
    for i in range(len(cxsmiles)):
        if cxsmiles[i] in ('H', '@', '[', ']', '(', ')', '=', '-', '#', ':', '+',
                           '1', '2', '3', '4', '5', '6', '7', '8', '9', '0', '/', '\\',
                           'l', 'r'): # 'l' for Cl, 'r' for 'Br'
            # print(cxsmiles[i])
            continue
        elif cxsmiles[i] == '*':
            if r_group_idx >= len(r_groups):
                raise Exception(f'Erorr in converting SMILES to CXSMILES, {smi}')
            # print(r_groups[r_group_idx])
            pos.append(f"{r_groups[r_group_idx]}")
            r_group_idx += 1
        else:
            pos.append('')
    pos = '|$' + ';'.join(pos) + '$|'
    return f'{cxsmiles} {pos}'

smi = get_cxsmiles_from_smi('[*:_R1]N1CCC[C@H]1C([*:_R2])=O')
print(smi)
display(Chem.MolFromSmiles(smi))


# clean dummy labels in cxsmiles
def clean_dummy_labels_in_cxsmiles(smi):
    """
    Clean dummy labels in cxsmiles
    Input: '*C(=O)[C@@H]1CCCN1C(C)=O |$_R2;;;;;;;;;;$,atomProp:0.dummyLabel.*|'
    Output: '[*]C(=O)[C@@H]1CCCN1C(C)=O |$_R2;;;;;;;;;;$|'
    """
    smi_parts = smi.split('|')
    smi = smi_parts[0].replace('*', '[*]') + '|$' + smi_parts[1].split('$')[1] + '$|'
    return smi

clean_dummy_labels_in_cxsmiles('*C(=O)[C@@H]1CCCN1C(C)=O |$_R2;;;;;;;;;;$,atomProp:0.dummyLabel.*|')



# Get SMI from CXSMILES
def get_smi_from_cxsmiles(cxsmiles):
    """
    Get SMILES from CXSMILES
    Input: 'CCCC[C@H](N(C)[*])C([*])=O |$;;;;;;;_R1;;_R2;$|'
    Output: 'CCCC[C@H](N(C)[*:_R1])C([*:_R2])=O'
    """
    smi_list = cxsmiles.split('|')
    smi, pos = smi_list[0], smi_list[1]
    labels = pos[1:-1].split(';')

    # Replace * with *:label according to the occurence of * in the SMILES
    for label in labels:  
        if len(label) == 0:
            continue
        index = smi.index('[*]')
        smi = smi[:index] + f'[*:{label}]' + smi[index+3:]

    return  smi.strip()

smi = get_smi_from_cxsmiles('CCCC[C@H](N(C)[*])C([*])=O |$;;;;;;;_R1;;_R2;$|')
print(smi)
display(Chem.MolFromSmiles(r_group_label2index(smi)))
```

**Reply by charlesxu90** [reply]
The input is like below:
![image](https://user-images.githubusercontent.com/5413227/229310865-b4e163c7-078a-4a62-9b0c-9db437f0a2ca.png)

```python
smi = '[*:_R3]O[C@H](C)[C@H](NC(=O)[C@@H](CC(C)C)N(C)C(=O)[C@@H]1CCCN1C(C)=O)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N(C)[C@@H](C)C(=O)N[C@H](CC(C)C)C(=O)N1CCC[C@@H]1C([*:_R2])=O'
Chem.MolFromSmiles(get_cxsmiles_from_smi(smi))
```

The ideal output is as follow:
![image](https://user-images.githubusercontent.com/5413227/229311300-94f8155d-56c5-42c7-b982-6121e440ef98.png)

```python
Chem.MolFromSmiles('CC(=O)N1CCC[C@H]1C(=O)N(C)[C@H](CC(C)C)C(=O)N[C@@H]1C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N(C)[C@@H](C)C(=O)N[C@H](CC(C)C)C(=O)N2CCC[C@@H]2C(=O)O[C@@H]1C')
```

**Reply by charlesxu90** [reply]
@greglandrum If there're no direct way to do it,  then I can inplement one myself. Could you give me some related implementations? 

Is it good to add this implementation into `Chem.molzip`?  Or there're some other functions that is better suited to be updated to solve this problem?

**Reply by bp-kelley** [reply]
molzip could easily add an option to cyclize structures, but we would need to supply the mapping.  The code to do this, however, isn't terribly hard.  For your example, I've used dummy atom labels as opposed to _R2, _R3

```
def get_nbr_atom(atom):
    """Get the nbr atom (must only be one, used to find dummy atom connections)"""
    nbrs = []
    for bond in atom.GetBonds():
        if bond.GetBeginAtomIdx() != atom.GetIdx():
            nbrs.append(bond.GetBeginAtom())
        else:
            nbrs.append(bond.GetEndAtom())
    assert len(nbrs) == 1
    return nbrs[0]
```

```
smi = '[*:3]O[C@H](C)[C@H](NC(=O)[C@@H](CC(C)C)N(C)C(=O)[C@@H]1CCCN1C(C)=O)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N(C)[C@@H](C)C(=O)N[C@H](CC(C)C)C(=O)N1CCC[C@@H]1C([*:2])=O'

# make a read write molecule
m = Chem.RWMol(Chem.MolFromSmiles(smi))

# find the atoms to connect
mapped_atoms = [atom for atom in m.GetAtoms() if atom.GetAtomMapNum()]
atoms_to_bond = {atom.GetAtomMapNum():get_nbr_atom(atom) for atom in mapped_atoms}

# bond them together, you have to know that :3 maps to :2, I don't know how to do this automatically
m.AddBond(atoms_to_bond[2].GetIdx(), atoms_to_bond[3].GetIdx(), Chem.BondType.SINGLE)

# remove the dummy atoms
m.BeginBatchEdit()
for atom in mapped_atoms:
    m.RemoveAtom(atom.GetIdx())
m.CommitBatchEdit()
```

![image](https://user-images.githubusercontent.com/9888416/229795675-703c3eac-9fb7-4fd3-bc33-1a5df91f393a.png)

We could add a special case to molzip to allow bonding, here is simple smiles examples below

```
molzip(smi, "[*:2][:*3]")
```

This currently will fail as the dummy atoms have nothing between them, but that is fixable in the code.

**Reply by greglandrum** [reply]
This doesn't strike me as something that really fits within what molzip is for.
It's easy enough to solve using normal RDKit code... sample code coming

**Reply by greglandrum** [reply, marked as answer]
Starting from a smiles like this (which is just what you have with the `_R` bits removed):
```
smi = '[*:3]O[C@H](C)[C@H](NC(=O)[C@@H](CC(C)C)N(C)C(=O)[C@@H]1CCCN1C(C)=O)C(=O)N(C)[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N[C@@H](CC(C)C)C(=O)N(C)[C@@H](C)C(=O)N[C@H](CC(C)C)C(=O)N1CCC[C@@H]1C([*:2])=O'
```
This function does what I think you're looking for:
```
def connect_mapped_atoms(mol,end1,end2):
    ''' end1 and end2 are the atom map numbers (integers) to look for
    
    note that this will not get atomic stereochemistry right
    '''
    res = Chem.RWMol(mol)
    dummy1 = None
    dummy2 = None
    for atom in res.GetAtoms():
        if atom.GetAtomMapNum()==end1:
            dummy1 = atom
        elif atom.GetAtomMapNum()==end2:
            dummy2 = atom
    assert dummy1 is not None and dummy2 is not None
    assert dummy1.GetDegree()==1
    assert dummy2.GetDegree()==1
    nbr1 = dummy1.GetNeighbors()[0]
    nbr2 = dummy2.GetNeighbors()[0]
    res.BeginBatchEdit()
    res.RemoveAtom(dummy1.GetIdx())
    res.RemoveAtom(dummy2.GetIdx())
    res.AddBond(nbr1.GetIdx(),nbr2.GetIdx(),Chem.BondType.SINGLE)
    res.CommitBatchEdit()
    return res
```

**Reply by charlesxu90** [reply]
@greglandrum @bp-kelley  Thanks a lot for the solution. I'm implementing it and find some other issues.

I'll post my final solution once it works.

Again, thanks for your attention and time. Best!

**Reply by charlesxu90** [reply]
The solution @greglandrum  works well for my problem.

During check, I found the conversion of smiles to cxsmiles has some errors. It's now fixed in the code commented above.
